File: services/core-api/app/schemas/operations.py
from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, ConfigDict


class OperationSchema(BaseModel):
    operation_id: str
    created_at: datetime
    operation_type: str
    status: str

    merchant_id: str
    terminal_id: str
    client_id: str
    card_id: str

    amount: int
    currency: str = "RUB"
    captured_amount: int = 0
    refunded_amount: int = 0

    daily_limit: Optional[int] = None
    limit_per_tx: Optional[int] = None
    used_today: Optional[int] = None
    new_used_today: Optional[int] = None

    authorized: bool = False
    response_code: str = "00"
    response_message: str = "OK"

    parent_operation_id: Optional[str] = None
    reason: Optional[str] = None

    mcc: Optional[str] = None
    product_code: Optional[str] = None
    product_category: Optional[str] = None
    tx_type: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


class OperationsPage(BaseModel):
    items: List[OperationSchema]
    total: int
    limit: int
    offset: int


class OperationTimeline(BaseModel):
    root: OperationSchema
    children: List[OperationSchema]


# Удобный псевдоним для журнала транзакций
TransactionLogPage = OperationsPage

File: frontends/admin-ui/src/types/operations.ts
export type OperationType = "AUTH" | "CAPTURE" | "REFUND" | "REVERSAL";
export type OperationStatus = string;

export interface Operation {
  operation_id: string;
  created_at: string;
  operation_type: OperationType;
  status: OperationStatus;
  merchant_id?: string | null;
  terminal_id?: string | null;
  client_id?: string | null;
  card_id?: string | null;
  amount: number;
  currency: string;
  captured_amount: number;
  refunded_amount: number;
  parent_operation_id?: string | null;
  product_code?: string | null;
  product_category?: string | null;
  tx_type?: string | null;
  mcc?: string | null;
  risk_result?: string | null;
  risk_score?: number | null;
  risk_reasons?: string[] | null;
  risk_flags?: Record<string, unknown> | null;
  risk_source?: string | null;
  risk_level?: string | null;
  risk_rules_fired?: string[] | null;
  risk_payload?: RiskPayload | null;
  daily_limit?: number | null;
  limit_per_tx?: number | null;
  used_today?: number | null;
  new_used_today?: number | null;
  authorized?: boolean | null;
  response_code?: string | null;
  response_message?: string | null;
  reason?: string | null;
}

export interface OperationListResponse {
  items: Operation[];
  total: number;
  limit: number;
  offset: number;
}

export interface OperationQuery extends Record<string, unknown> {
  limit?: number;
  offset?: number;
  operation_type?: string;
  status?: string;
  merchant_id?: string;
  terminal_id?: string;
  client_id?: string;
  card_id?: string;
  date_from?: string;
  date_to?: string;
  from_created_at?: string;
  to_created_at?: string;
  min_amount?: number;
  max_amount?: number;
  risk_result?: string | string[];
  risk_min_score?: number;
  risk_max_score?: number;
  order_by?:
    | "created_at_desc"
    | "created_at_asc"
    | "amount_desc"
    | "amount_asc"
    | "risk_score_desc"
    | "risk_score_asc";
  risk_level?: string | string[];
  mcc?: string;
  product_category?: string;
  tx_type?: string;
}

export interface RiskDecisionPayload {
  level?: string;
  reason_codes?: string[];
  rules_fired?: string[];
  ai_score?: number;
  ai_model_version?: string;
}

export interface RiskPayload {
  decision?: RiskDecisionPayload;
  source?: string;
  flags?: Record<string, unknown>;
  [key: string]: unknown;
}
